### DISCUSSION TITLE
Running all tests of mono repo

### DISCUSSION THREAD

**Discussion by Ohad31415** [pytest-dev/pytest#9504] [Q&A] [answered]
We have one repo nesting several projects, something like this - 
<pre>
.
├── project1
│   ├── src/
│   └── tests/
│      ├── __init__.py
│      ├── ...
├── project2
│   ├── src/
│   └── tests/
│      ├── __init__.py
│      ├── ...
├──  ...
</pre>

Now we run tests for each project separately from its directory and if I try to `python -m pytest` from the repo root I get an import errors. 
Is there a way to do have he ability to run all the tests of all those project in one command? 🤔

**Comment by nicoddemus** [comment, marked as answer]
Hi @Ohad31415,

There's nothing builtin in pytest, I'm afraid.

**Reply by Ohad31415** [reply]
Ok @nicoddemus thanks for answer at least I'll clear this thread from my mind 😁. 
It will be a nice feature maybe being able to inform pytest (like mentioning path to each project) of such structure so it can collect and run tests for each project sequentially.

**Reply by nicoddemus** [reply]
> Ok @nicoddemus thanks for answer at least I'll clear this thread from my mind 😁.

Cool!

> It will be a nice feature maybe being able to inform pytest (like mentioning path to each project) of such structure so it can collect and run tests for each project sequentially.

Not sure, because pytest works like most (all?) tools in that regard. For example, if we go down this path, should other tools (like `black`, or `coverage`, for example) follow suit? I don't think so, specially because this is easy to implement with a custom shell script in the root of the monorepo, and would solve this not only for running pytest, but for every other task needed (say compilation if you have C extensions).

Having said that, this can probably be implemented as a pytest plugin, if one really wants to pursue this. 😁

**Reply by Ohad31415** [reply]
@nicoddemus Well yeah you're right, I guess our current approach is not too bad and yeah is not something too difficult.. 🙂
Thanks again.

**Reply by LordFckHelmchen** [reply]
Hm... not really an answer, right (just saying because it shows up as answered - which it isn't).

The use case we have is a lib on project2 and an implementation that uses the lib in project1. Coverage however would only be correct when we measure tests of project1 using project2 and not the projects in isolation. Therefore we need to run all together.

**Comment by adriangb** [comment]
Does it help at all if the repo structure is:

```
.
├── project1
│   └── src/
│      ├── __init__.py
│      ├── tests/
|      |   └── __init__.py
│      ├── ...
├── project2
│   └── src/
│      ├── __init__.py
│      ├── tests/
|      |   └── __init__.py
│      ├── ...
```

That is `project1.tests` and `project2.tests` are valid modules. I've found the `project1/tests` structure to be problematic because it causes conflicting `tests` modules, but the `project1/src/tests` structure solves that. But running `pytest .` still gives me import errors.

**Reply by LordFckHelmchen** [reply]
That will include all your tests in your built packages though... if you are not careful ;-)

**Comment by dimaqq** [comment]
I'm trying to wrap my head about this, and I think it comes down to:

- `project1` is not a package
- `project2` is not a package
- (also, `src` is not a package, but that's irrelevant)

Pytest tries to import `tests.test_foo` after discovering `tests/test_foo.py` and from the standpoint of Python's import machinery, `tests.` can only refer to a single directory, not both.

**Comment by xixixao** [comment]
With uv supporting monorepo workspaces, it would be nice for pytest to support this as well. We're running a for loop over all projects, but that doesn't give a single overview of how many tests pass or fail.
